### PR TITLE
feat(532): Token model

### PR DESCRIPTION
**Pull request by ian-fox**
Add validation for tokens and their endpoints, and tests for the validation.

Currently the tokens themselves are not returned in the GET request, I figure for the one time we display them they can probably be returned to the request that was made to create the new one.

### CODE REVIEW

File: models/token.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+     *
+     * @property update
+     * @type {Joi}
+     */
+    update: Joi.object(mutate(MODEL, [
+        'id'
+    ], [
+        'name',
+        'description',
+        'lastUsed'
```
Comment (FenrirUnbound) [inline models/token.js]: The `lastUsed` attribute is intended to describe when the token was last accessed. Why do we want the caller to be able to update it?

Comment (ian-fox) [inline models/token.js]: Oversight on my part

**Comment by ian-fox** [comment]
@FenrirUnbound any other changes?

File: models/token.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    token: Joi
+        .string()
+        .token(),
+
+    id: Joi
+        .number()
+        .integer()
+        .positive(),
+
+    user: Joi
```
Comment (d2lam) [inline models/token.js]: `userId` would be clearer

File: models/token.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        .integer()
+        .positive(),
+
+    userId: Joi
+        .number()
+        .integer()
+        .positive()
+        .description('User ID'),
+
+    name: Joi
```
Comment (petey) [inline models/token.js]: Token names should probably be very similar to secret names.

https://github.com/screwdriver-cd/data-schema/blob/master/config/job.js#L23

Comment (ian-fox) [inline models/token.js]: So like SCHEMA_TOKEN, SCHEMA_ID, SCHEMA_USER_ID etc?

Comment (FenrirUnbound) [inline models/token.js]: Secret names are used by the build systems, so I can understand why they follow certain conventions. Token names are for the human users to better differentiate between various secrets; the system won't care what they're named. Why would we specifically want token names to be [all uppercase and without spaces](https://github.com/screwdriver-cd/data-schema/blob/master/config/regex.js#L21)

File: models/token.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    ])).label('Create token'),
+
+    /**
+     * Properties for token that will be passed during a UPDATE requeste
+     *
+     * @property update
+     * @type {Joi}
+     */
+    update: Joi.object(mutate(MODEL, [
+        'id'
```
Comment (minz1027) [inline models/token.js]: We won't need `id` here, if we are trying have a update endpoint like this: PUT /tokens/{id}.
https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/secrets/update.js
The `id` shouldn't be in the payload.

Comment (ian-fox) [inline models/token.js]: Ooh, right. Not sure what I was thinking there, thanks!

File: models/token.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        .example('Mobile token'),
+
+    description: Joi
+        .string()
+        .max(256)
+        .description('Token description')
+        .example('Used to authenticate the mobile app'),
+
+    lastUsed: Joi
+        .date()
```
Comment (petey) [inline models/token.js]: This should probably be `.string().isoDate()` to be consistent with other timestamps.
https://github.com/screwdriver-cd/data-schema/blob/master/models/build.js#L21-L25

File: models/token.js (hunk 3 above)
Comment (petey) [inline models/token.js]: ok, I'm fine with that.

File: models/token.js (hunk 5 above)
Comment (FenrirUnbound) [inline models/token.js]: use `isoDate` to keep this consistent with the other time-based attributes that other model types have. 

https://github.com/screwdriver-cd/data-schema/blob/master/models/build.js#L23

Comment (ian-fox) [inline models/token.js]: Done!

Comment (ian-fox) [inline models/token.js]: Done!

**Comment by ian-fox** [comment]
@petey @FenrirUnbound Made one more modification upon testing and realizing that every model needs an allKeys property. Sorry for the last minute change!

**Pull request merged**